File: triad-core/src/main/java/com/nhaarman/triad/DefaultTransitionAnimator.java
package com.nhaarman.triad;

import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import com.nhaarman.triad.Triad.Callback;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

class DefaultTransitionAnimator implements TransitionAnimator {

    public static final DefaultTransitionAnimator INSTANCE = new DefaultTransitionAnimator();

    private DefaultTransitionAnimator() {
    }

    @Override
    public boolean forward(@Nullable final View oldView, @NotNull final View newView, @NotNull final ViewGroup parent, @NotNull final Callback callback) {
        executeAlphaTransition(oldView, newView, parent, callback);
        return true;
    }

    @Override
    public boolean backward(@Nullable final View oldView, @NotNull final View newView, @NotNull final ViewGroup parent, @NotNull final Callback callback) {
        executeAlphaTransition(oldView, newView, parent, callback);
        return true;
    }

    private void executeAlphaTransition(
          @Nullable final View oldView,
          @NonNull final View newView,
          @NonNull final ViewGroup parent,
          @NonNull final Callback callback
    ) {
        int durationMs = parent.getContext().getResources().getInteger(android.R.integer.config_shortAnimTime);

        parent.addView(newView);

        if (oldView != null) {
            oldView.animate()
                  .alpha(0f)
                  .setDuration(durationMs);
        }

        newView.setAlpha(0f);
        newView.animate()
              .alpha(1f)
              .setDuration(durationMs)
              .withEndAction(new Runnable() {
                  @Override
                  public void run() {
                      if (oldView != null) {
                          parent.removeView(oldView);
                      }

                      callback.onComplete();
                  }
              });
    }
}

File: triad-core/src/main/kotlin/com/nhaarman/triad/DefaultTransitionAnimator.kt
package com.nhaarman.triad

import android.view.View
import android.view.ViewGroup

internal object DefaultTransitionAnimator : TransitionAnimator {

    override fun forward(oldView: View?, newView: View, parent: ViewGroup, onComplete: () -> Unit): Boolean {
        executeAlphaTransition(oldView, newView, parent, onComplete)
        return true
    }

    override fun backward(oldView: View?, newView: View, parent: ViewGroup, onComplete: () -> Unit): Boolean {
        executeAlphaTransition(oldView, newView, parent, onComplete)
        return true
    }

    private fun executeAlphaTransition(oldView: View?, newView: View, parent: ViewGroup, onComplete: () -> Unit) {
        val durationMs = parent.context.resources.getInteger(android.R.integer.config_shortAnimTime).toLong()

        parent.addView(newView)

        oldView?.animate()
              ?.alpha(0f)
              ?.setDuration(durationMs)

        newView.alpha = 0f
        newView.animate()
              .alpha(1f)
              .setDuration(durationMs)
              .withEndAction {
                  if (oldView != null) {
                      parent.removeView(oldView)
                  }

                  onComplete()
              }
    }
}
